### PR TITLE
Support a closure for instantiating RedirectMiddleware

### LINKED ISSUE #2363: Remember original request location from Auth Redirect Middleware
I'm not sure if this is an enhancement request or a question (aka. the feature already exists). 

Currently the redirect middleware is a great options to redirect away from protected routes while unauthenticated. However, it does not remembers the location of the original request after redirecting. So if someone is taken to the login page from a direct link (e.g. from an email), they will not be able to go back to the original link after logging in. 

I think vapor should add a feature that allows retrieving the original link before redirecting using the redirect middleware (or maybe this feature already exists?). 

### Environment

* Vapor Framework version: Vapor 4.5.1
* OS version: macOS 10.15.4

### PR DESCRIPTION
**Pull request by TomShen1234**
`RedirectMiddleware` now supports a closure with a `Request` object where you can setup a custom redirect based on the original request. (#2364, fixes #2363)

```swift
Authenticatable.redirectMiddleware(makePath: { req in
    "/login?orig=\(req.url.path)"
})
```

### CODE REVIEW

**Comment by TomShen1234** [comment]
@Joannis Thanks for the feedback. This is only a rough draft. Do you think it would be better if the entire `Request` was passed back to the closure in order to create the new dynamic path? This would be an simple change to make since the middleware also have access to the original `Request`. This would give the closure access to the request's header and other metadata. I can modify with another commit if you think this idea is more feasible.

Currently I don't think there is a way to provide any kind of dynamic redirect with the current redirect middleware which only supports a static `String` when grouping the middlewares.

File: Sources/Vapor/Authentication/RedirectMiddleware.swift [hunk 1]
```diff
@@ -6,24 +6,42 @@ extension Authenticatable {
...
 
 private final class RedirectMiddleware<A>: Middleware
     where A: Authenticatable
 {
-    let path: String
+    let pathClosure: PathClosure
+    
+    typealias PathClosure = (Request) -> String
 
     init(_ authenticatableType: A.Type = A.self, path: String) {
```
Comment (tanner0101) [inline Sources/Vapor/Authentication/RedirectMiddleware.swift:28]: This initializer is internal so we can delete it, no need for two. As long as we keep `public static func redirectMiddleware(path: String) -> Middleware` we're fine.

File: Sources/Vapor/Authentication/RedirectMiddleware.swift [hunk 2]
```diff
@@ -6,24 +6,42 @@ extension Authenticatable {
...
 }
 
 
 private final class RedirectMiddleware<A>: Middleware
     where A: Authenticatable
 {
-    let path: String
+    let pathClosure: PathClosure
+    
+    typealias PathClosure = (Request) -> String
```
Comment (tanner0101) [inline Sources/Vapor/Authentication/RedirectMiddleware.swift:26]: -1 to closure aliases, it's an extra layer of indirection to juggle when reading code.

File: Sources/Vapor/Authentication/RedirectMiddleware.swift [hunk 3]
```diff
@@ -6,24 +6,42 @@ extension Authenticatable {
...
     }
 
     /// See Middleware.respond
     func respond(to req: Request, chainingTo next: Responder) -> EventLoopFuture<Response> {
         if req.auth.has(A.self) {
             return next.respond(to: req)
         }
-        let redirect = req.redirect(to: path)
+        
+        let redirectPath = pathClosure(req)
```
Comment (tanner0101) [inline Sources/Vapor/Authentication/RedirectMiddleware.swift:42]: Missing `self.`

File: Sources/Vapor/Authentication/RedirectMiddleware.swift [hunk 4]
```diff
@@ -6,24 +6,42 @@ extension Authenticatable {
     public static func redirectMiddleware(path: String) -> Middleware {
         return RedirectMiddleware<Self>(Self.self, path: path)
     }
+    
+    /// Basic middleware to redirect unauthenticated requests to the supplied path
+    ///
+    /// - parameters:
+    ///    - pathClosure: The closure that returns the redirect path based on the given URI path
+    ///    - input: The original `Request` object of the request
+    public static func redirectMiddleware(pathClosure: @escaping (_ input: Request) -> String) -> Middleware {
```
Comment (tanner0101) [inline Sources/Vapor/Authentication/RedirectMiddleware.swift:15]: I think `makePath`, `pathFactory`, or just `_` would be better names here. The closure definition doesn't need a label or doc block entry either, that can be just `(Request) -> String`.

**Comment by tanner0101** [comment]
Please update PR body / title for release notes. Furthermore, a small PR adding a note to https://github.com/vapor/docs/blob/master/4.0/docs/authentication.md explaining how to use this would be great.

**Comment by TomShen1234** [comment]
@tanner0101 Made the requested modifications and updated the PR body. I'll start working on updating the docs.

**Comment by 0xTim** [comment]
This will need tests as well

**Comment by TomShen1234** [comment]
@0xTim @tanner0101 What do you want the tests to look like? I didn't find an existing test in the tests folder for this middleware that I can modify (or am I understanding this wrong?).

**Comment by 0xTim** [comment]
@TomShen1234 it's highly likely there are no existing tests to use. But there should be _some_ other tests for other middleware. Basically just a test that ensures the closure gets called if supplied would be great

**Comment by TomShen1234** [comment]
I have added a new test in `AuthenticationTests.swift`. It's based on the `testBasicAuthenticator()`, except instead of using a guard middleware, it uses the redirect middleware with the new closure. Please review, thanks!

**Comment by tanner0101** [comment]
These changes are now available in [4.15.0](https://github.com/vapor/vapor/releases/tag/4.15.0)

**Pull request merged**